### DISCUSSION TITLE
Spacy cli pretrain function

### DISCUSSION THREAD

**Discussion by MartiLevizzani** [explosion/spaCy#11007] [Help: Coding & Implementations]
Good morning,
I would like to understand how to correctly set the pretrain function offered by spacy v3.

I am using spacy on Google Colab, saving the various folders with the models on google drive (from which I extract the paths). I would like to retrain some costum models on new prepared datasaets, starting from the one already pre-trained in order to have an improvement of the model.

For the train I used the train cli function correctly, but I have difficulty in setting the pretrain parameters, first of all can my problem be solved with this function?

Could I know more fully what the fields to be entered refer to? (ex: the correct path of the resume-path)

The configuration file, from the documentation, is the same as the train, but it reports errors to me

Thank you so much
Martina

- [![image](https://user-images.githubusercontent.com/15217151/175053619-7ff47094-3f3b-4430-9776-1417ae8822e5.png) ]

**Comment by polm** [comment]
Hey, please do not past screenshots of code or errors, copy and paste them as text.

**Comment by polm** [comment]
First, to be clear, you don't have to use pretraining, it's an optional feature. In some situations it can help, but not all, so it's fine not to use it. If you are just getting started I would suggest not using until you get your models working generally.

The values for arguments for pretrain are [covered in the docs](https://spacy.io/api/cli/#pretrain). Also for pretraining, while you can use the same config file as for normal training, you need to add more fields to it. This is covered in the [usage docs on pretraining](https://spacy.io/usage/embeddings-transformers/#pretraining). Normally you would not call pretrain from the function, you would instead call it from the command line. You can use the command line even in Google Colab. Is there a reason you are using the function instead?

You didn't share any error, so I'm not sure what's not working. What is the error you got? Did you put the pretrain fields in your config?

**Reply by MartiLevizzani** [reply]
I generate the config file in that way

!python -m spacy init fill-config "/content/drive/MyDrive/BILANCI/analisi spacy/test - pretrain/config.cfg" "/content/drive/MyDrive/BILANCI/analisi spacy/test - pretrain/config_pretrain.cfg" --pretraining

and all the field are correctly compiled
but when I call pretrain by cli in that way: 
pretrain(config='/content/drive/MyDrive/BILANCI/Bilanci esercizio SPA/spacy/sicurezza/config.cfg', output_dir= "/content/drive/MyDrive/BILANCI/analisi spacy/prova/", resume_path='/content/drive/MyDrive/BILANCI/Bilanci esercizio SPA/spacy/experiment/v2/model-best', epoch_resume=None, use_gpu=0, silent=False)

this is the error:

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-4-842d4abd6312>](https://localhost:8080/#) in <module>()
----> 1 pretrain(config='/content/drive/MyDrive/BILANCI/Bilanci esercizio SPA/spacy/sicurezza/config.cfg', output_dir= "/content/drive/MyDrive/BILANCI/analisi spacy/prova/", resume_path='/content/drive/MyDrive/BILANCI/Bilanci esercizio SPA/spacy/experiment/v2/model-best', epoch_resume=None, use_gpu=0, silent=False)

[/usr/local/lib/python3.7/dist-packages/spacy/training/pretrain.py](https://localhost:8080/#) in pretrain(config, output_dir, resume_path, epoch_resume, use_gpu, silent)
     27 ):
     28     msg = Printer(no_print=silent)
---> 29     if config["training"]["seed"] is not None:
     30         fix_random_seed(config["training"]["seed"])
     31     allocator = config["training"]["gpu_allocator"]

TypeError: string indices must be integers

**Reply by adrianeboyd** [reply]
This is not the same `pretrain` function, and this function is expecting a `Config` object, not a path to a config file. (The CLI is much easier to use!)
